Adds Homebrew and completion support to asdf plugin

jburwell opened the pull request:
* Modifies the search logic for asdf to include Homebrew when it is
installed.  The implementation is adapted from the pyenv plugin.

* Includes command completions for both regular and Homebrew installs.
This enhancement was inspired by PR #6046.

mcornella commented:
For what I understand, the only places that asdf can be in is either `$ASDF_DIR` or in whatever path Homebrew installs it, so there's no need to check all these directories: just checking the Homebrew path if the other wasn't found should be enough.

Also, if you installed it with homebrew you can probably get what directory is installed in using `$commands[asdf.sh]` or something similar. But that depends on how Homebrew makes the install. What does `brew --prefix asdf` return?

jburwell commented:
@mcornella my original thinking around the search path was to follow pyenv's approach.  Based on your feedback, I realized that these additional search paths were present to pyenv when it may be installed as part of distribution packages.  Since asdf does not appear to be packaged by any distros, this search logic is unnecessary.  Therefore, I have updated the patch to search ``$HOME/.asdf`` and the Homebrew location (if Homebrew is installed).

Also, as part of this update, I extracted the common bits of asdf startup into a common code block.

Finally, ``brew --prefix asdf`` provides the root directory of the asdf installation under Homebrew.  This approach avoid hard coding the root path or assuming the directory layout used by the asdf Homebrew package.

mcornella commented:
I've made a number of changes that simplify the logic, and also fixes the paths that you were looking for in the case of Homebrew installations.
